### ISSUE TITLE
Question on using different null models

### ISSUE THREAD

**Issue by dannypanduro** [GenLouvain/GenLouvain#6]
Hello,

This isn't an issue, I have a question that I haven't figured been able to figure out. I apologize for writing in the issues tab. 

I am new to Matlab and coding, and I'm having difficulty figuring out how to apply a 'Geographical' Null model to the Louvain algorithm. The geographical null model has been described in much of Dr. Danielle Bassett's works with community detection.
Pij = pBij  
Pij is the null model, p is the mean edge weight, and Bij is the binary adjacency matrix of the network. 

I saw a comment in the modularity function that the function could be altered for different null models; however, I am unsure how to do this. I would truly appreciate any guidance on how to incorporate the formula into the code.

Sincerely, 
Richard

**Comment by LJeub** [comment]
Hi,

it seems the 'geographical' null model in Danielle Bassett's work is also known as the constant Potts model. In this case you can optimize this by using
```matlab
    S = genlouvain(full(A-p))
```
where `A` is the (weighted) adjacency matrix and `p` is the mean edge weight, i.e., `p=mean(A(:))`.   Note that this is different from what your formula suggests. Further, you can change the resolution by changing `p`.

If you actually only want to apply the null model to the non-zero edges as your formula suggests, you could do this by using 
```matlab
   S = genlouvain(full(A-p*logical(A)))
```
with `p=mean(A(A>0))`. However, from a quick check, this seems to be different from what is done in Danielle Bassett's work.

Note the use of `full` here is to make sure you have a dense matrix as input in case your adjacency matrix `A` is in sparse format which should give you better performance. Also check out `help(genlouvain)` to have a look at the different randomization options (by default the only randomization is shuffling the order of nodes) and make sure the features you observe are robust.

**Comment by dannypanduro** [comment]
Hello,

I'm sorry for the late reply, I spent yesterday sorting through and attempting to visualize the formulas you mentioned. Thank you for replying, this has been very educational. 

If I wanted the nodes only to have edges with their immediate neighbors, does `S = genlouvain(full(A-p*logical(A)))` work?
Also, is the formula above different from the formula Danielle Bassett uses in her [2015 paper](https://www.math.ucla.edu/~mason/papers/softmatter-final.pdf) because it doesn't contain a Kronecker delta verifying if the nodes are in the same community? 
![Basset et al , 2015](https://user-images.githubusercontent.com/81585331/126534773-aeec4c45-5e4d-4b5e-b9e3-c86a5d22731a.png)

Thank you for your insight,
Richard

**Comment by LJeub** [comment]
Hi,

yes, this should compute the modularity with geographic null-model from that paper. The summation with the Kronecker delta is computed internally in `genlouvain` as it moves nodes to different communities. The input to the algorithm is only the matrix with entries
`W_{ij} - \gamma P_{ij}`, often called the modularity matrix.

**Comment by NEUZhangGH** [comment]
Hello,

I used the same geographical_null_model, but the project can work only my data is few. My system to be detected is so large that the project goes wrong. How should I do? The errors happen here:"Line 212 n= Length (B)". I changed it into "[n,nnn]=size(B)" and it goes on. Now it's error is "Line 226  it(:,1)=M(1)". I tried to change it into "it(:,1)=M(:,1)" but it reminds me cannot use ":".
However these errors don't happen when I test using few data. So I think if I wrongly used the project, or there is another way to set the function. My function now is:"A=@(i)W(:,i)-gamma*B(:,i);" . I am so confused and I beg for your help! 

Sincerely,
Zhang
